### ISSUE TITLE
SPLOM diagonals

### ISSUE THREAD

**Issue by nicolaskruchten** [plotly/plotly_express#42]
from https://community.plot.ly/t/plotly-express-scatterplot-matrix-diagonal/21464

> Is it somehow possible to set the diagonal of the scatterplot matrix in plotly express to a histogram like in pandas? And even more better like seaborns pairgrid a stack histogram with its argument hue? Would be really nice.

**Comment by alexcjohnson** [comment]
Would love this to be built in at the plotly.js level. From https://github.com/plotly/plotly.js/issues/2555:

> ### Things we could do, but would require some thought
> * Add _sub-trace_ on the diagonal (e.g. with `diagonal.mode: 'scattergl' || 'histogram' || 'box' || 'violin'`, see [#2505 (comment)](https://github.com/plotly/plotly.js/pull/2505#discussion_r177823962)

**Comment by nicolaskruchten** [comment]
Yeah we definitely need this baked in at the JS level! I'd want brushing to match, meaning if I lasso-select some points, I'd see "part of" the diagonal being selected somehow. Easy with histograms (just shade part of the bar) but harded with box/violin.

I'd also love to be able to have contours in the upper triangle and points in the lower triangle, or something like that. Brushing in that case seems very challenging though:

![image](https://user-images.githubusercontent.com/203523/55488870-d7145980-55fe-11e9-94e3-ed9fdb009a91.png)

**Comment by alexcjohnson** [comment]
Ah yeah, contours on one triangle would be cool! One easy way to handle brushing in that case would be to just use the coordinates of the outline to select the corresponding points as though the display was scatter. On the diagonal I guess brushing should be limited to 1D regardless of the display mode.

**Comment by nicolaskruchten** [comment]
Oh no I mean the other way: if you select some blob-shaped assortment of points in a scatter subplot, how do you render the contours/histogram/boxes?

**Comment by alexcjohnson** [comment]
Ah I see. For all the diagonal options I feel like we can make a separate "filtered" trace - so a second histogram inside the first, a separate violin/box next to the first.

Definitely tricky for contours though... it's hard to make two overlaid contour plots useful. So either we just plot contours of the selected data, or perhaps faded black & white contours of the full data set with semi-transparent contours in color on top of it, kind of like parcoords

**Comment by nicolaskruchten** [comment]
Something like that, yep

**Comment by wlievens** [comment]
This would be really neat, is anything like this planned?

**Comment by grisaitis** [comment]
Is this not possible with `graph_objects.Splom`?

**Comment by nicolaskruchten** [comment]
@grisaitis not yet! Plotly Express uses `graph_objects.Splom` under the hood so if it were possible there we'd have it in PX as well already :)

**Comment by grisaitis** [comment]
Ok so the roadblock is that the main dependency doesn't support this. Do we know of an open issue in plotly/plotly about this?

**Comment by nicolaskruchten** [comment]
Yep: https://github.com/plotly/plotly.js/issues/2555

**Comment by ChristianMichelsen** [comment]
Any news on this? Coming from a data science background, this seems like a very common plot that would be super useful for EDA!

**Comment by InCogNiTo124** [comment]
Bump for histograms on the diagonals

**Comment by dragoljub** [comment]
👍  Would love to see this too!

**Comment by rsgoodwin** [comment]
@nicolaskruchten - I really like your suggestion.  Would be very nice to have this as a feature.

**Comment by milanzmitrovic** [comment]
Hi all,

This would be wunderful to have in Plotly express. Charts like this are really workhorse when presenting classification algorithms.

Seaborn has similar algo implemented https://seaborn.pydata.org/examples/scatterplot_matrix.html . It is kind of shame for Plotly not to have sth like that already implemented.

I saw similar implementation in figure factory, but ff seems not to be maintained anymore, correct me if I am wrong?

**Comment by UpstatePedro** [comment]
Is there any update on this?

**Comment by hughjward** [comment]
While not using px it is possible with the figure factory, as a workaround until this is added to px:
`ff.create_scatterplotmatrix`

using the `diag='histogram'` option.
